crypto/hash/hmac_nss.c

```diff
@@ -0,0 +1,287 @@
+        return srtp_err_status_bad_param;
+    }
+
+    /* Initialize NSS equiv of NSS_NoDB_Init(NULL) */
+    nss = NSS_InitContext("", "", "", "", NULL,
+                          NSS_INIT_READONLY | NSS_INIT_NOCERTDB |
+                              NSS_INIT_NOMODDB | NSS_INIT_FORCEOPEN |
+                              NSS_INIT_OPTIMIZESPACE);
+    if (!nss) {
+        return srtp_err_status_auth_fail;
+    }
+
+    *a = (srtp_auth_t *)srtp_crypto_alloc(sizeof(srtp_auth_t));
+    if (*a == NULL) {
+        NSS_ShutdownContext(nss);
+        return srtp_err_status_alloc_fail;
+    }
+
+    hmac =
+        (srtp_hmac_nss_ctx_t *)srtp_crypto_alloc(sizeof(srtp_hmac_nss_ctx_t));
+    if (hmac == NULL) {
+        NSS_ShutdownContext(nss);
+        srtp_crypto_free(*a);
+        *a = NULL;
+        return srtp_err_status_alloc_fail;
+    }
+
+    hmac->nss = nss;
+    hmac->key = NULL;
+    hmac->ctx = NULL;
+
+    /* set pointers */
+    (*a)->state = hmac;
+    (*a)->type = &srtp_hmac;
+    (*a)->out_len = out_len;
+    (*a)->key_len = key_len;
+    (*a)->prefix_len = 0;
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_dealloc(srtp_auth_t *a)
+{
+    srtp_hmac_nss_ctx_t *hmac;
+
+    hmac = (srtp_hmac_nss_ctx_t *)a->state;
+    if (hmac) {
+        /* free any PK11 values that have been created */
+        if (hmac->key) {
+            PK11_FreeSymKey(hmac->key);
+            hmac->key = NULL;
+        }
+
+        if (hmac->ctx) {
+            PK11_DestroyContext(hmac->ctx, PR_TRUE);
+            hmac->ctx = NULL;
+        }
+
+        if (hmac->nss) {
+            NSS_ShutdownContext(hmac->nss);
+            hmac->nss = NULL;
+        }
+
+        /* zeroize everything */
+        octet_string_set_to_zero(hmac, sizeof(srtp_hmac_nss_ctx_t));
+        srtp_crypto_free(hmac);
+    }
+
+    /* free memory */
+    srtp_crypto_free(a);
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_start(void *statev)
+{
+    srtp_hmac_nss_ctx_t *hmac;
+    hmac = (srtp_hmac_nss_ctx_t *)statev;
+
+    if (PK11_DigestBegin(hmac->ctx) != SECSuccess) {
+        return srtp_err_status_auth_fail;
+    }
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_init(void *statev,
+                                        const uint8_t *key,
+                                        int key_len)
+{
+    srtp_hmac_nss_ctx_t *hmac;
+    hmac = (srtp_hmac_nss_ctx_t *)statev;
+    PK11SymKey *sym_key;
+    PK11Context *ctx;
+
+    if (hmac->ctx) {
+        PK11_DestroyContext(hmac->ctx, PR_TRUE);
+        hmac->ctx = NULL;
+    }
+
+    if (hmac->key) {
+        PK11_FreeSymKey(hmac->key);
+        hmac->key = NULL;
+    }
+
+    PK11SlotInfo *slot = PK11_GetBestSlot(CKM_SHA_1_HMAC, NULL);
+    if (!slot) {
+        return srtp_err_status_bad_param;
+    }
+
+    SECItem key_item = { siBuffer, (unsigned char *)key, key_len };
+    sym_key = PK11_ImportSymKey(slot, CKM_SHA_1_HMAC, PK11_OriginUnwrap,
```

Comment:
This stuff actually seems a little more useful to make common across the NSS modules.  It seems like you could have a function of the following form that would wrap these lines:

```
PK11SymKey* srtp_nss_import_key(CK_METHOD method, const uint8_t *key, size_t key_len);
```

If you take that together with startup / shutdown methods, maybe that's enough for another `.c/.h`?